aptos-move/framework/aptos-token/sources/bid.move

```diff
@@ -0,0 +1,281 @@
+module aptos_token::bid {
+    use aptos_token::listing::{Self, Listing};
+    use aptos_token::token::{Self, TokenId};
+    use std::guid::{Self, ID};
+    use std::signer;
+    use aptos_framework::coin::{Self, Coin};
+    use aptos_framework::timestamp;
+    use aptos_std::table::{Self, Table};
+
+    const ENO_SUFFICIENT_FUND: u64 = 1;
+    const ETOKEN_ID_NOT_MATCH: u64 = 2;
+    const ELISTING_EXPIRED: u64 = 3;
+    const ETOKEN_AMOUNT_NOT_MATCH: u64 = 4;
+    const EBID_NOT_EXIST: u64 = 5;
+    const ECANNOT_DRAW_FUND_BEFORE_EXPIRATION_TIME: u64 = 6;
+    const ELISTING_ID_NOT_MATCH: u64 = 7;
+
+    /// hold the bid info and coin at user account
+    /// ensure the coin cannot be withdrawn or transferred during the auction
+    struct Bid<phantom CoinType> has store {
+        bidder: address,
+        coin: Coin<CoinType>,
+        token_id: TokenId,
+        token_amount: u64,
+        lock_until_sec: u64, // coin can be extracted from the Bid only after lock time expires
+        listing_id: ID, // ensure the bid is only executed for the chosen listing_id
+    }
+
+    /// store all the bids by the user
+    struct BidRecords<phantom CoinType> has key {
+        records: Table<ID, Bid<CoinType>>,
+    }
+
+    public fun initialize_bid_records<CoinType>(bidder: &signer)  {
+        let owner_addr = signer::address_of(bidder);
+
+        if (!exists<BidRecords<CoinType>>(owner_addr)) {
+            move_to(
+                bidder,
+                BidRecords<CoinType> {
+                    records: table::new(),
+                }
+            );
+        };
+    }
+
+    /// withdraw the coin and store them in bid struct and return a global unique bid id
+    public fun bid<CoinType>(
+        bidder: &signer,
+        token_id: TokenId,
+        token_amount:u64,
+        withdraw_amount: u64,
+        lock_until_sec: u64,
+        listing_id: ID,
+    ): ID acquires BidRecords {
+        let bidder_address = signer::address_of(bidder);
+
+        assert!(coin::balance<CoinType>(bidder_address) >= withdraw_amount, ENO_SUFFICIENT_FUND);
+        // withdraw the coin and store them in escrow to ensure the fund is avaliable until expiration_sec
+        let coin = coin::withdraw<CoinType>(bidder, withdraw_amount);
+
+        let bid = Bid<CoinType> {
+            bidder: bidder_address,
+            coin,
+            token_id,
+            token_amount,
+            lock_until_sec,
+            listing_id,
+        };
+        let bid_id = create_bid_id(bidder);
+        initialize_bid_records<CoinType>(bidder);
+        let bid_records = &mut borrow_global_mut<BidRecords<CoinType>>(bidder_address).records;
+        table::add(bid_records, bid_id, bid);
+        bid_id
```

Comment:
not clear how the smart contract will track bids against a specific token

Reply:
> The code currently lacks a clear story. For example, it isn't clear to me how this is used to drive an auction, how one might build royalties, or add fees. I think if we want to go in this direction, maybe the better choice is to construct a pretty comprehensive market place derived from the work here and #1204 and create something exploratory in `move-examples` until we land on a solution that at least demonstrates a clear end-to-end. From there, we can decide do we want to be opinionated or break it down into composable pieces.

will add an example to this PR to show the e2e flow.